cmd/dagger/query.go

```diff
@@ -0,0 +1,204 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"dagger.io/dagger"
+	"github.com/spf13/cobra"
+	"golang.org/x/term"
+)
+
+var (
+	queryFile      string
+	queryVarsInput map[string]string
+)
+
+var queryCmd = &cobra.Command{
+	Use:                   "query [flags] [operation]",
+	Aliases:               []string{"q"},
+	DisableFlagsInUseLine: true,
+	Long:                  "Send API queries to a dagger engine\n\nWhen no document file, read query from standard input.",
+	Short:                 "Send API queries to a dagger engine",
+	Example: `
+dagger query <<EOF
+{
+  container {
+    from(address:"hello-world") {
+      exec(args:["/hello"]) {
+        stdout {
+          contents
+        }
+      }
+    }
+  }
+}
+EOF
+`,
+	Run:  Query,
+	Args: cobra.MaximumNArgs(1), // operation can be specified
+}
+
+func Query(cmd *cobra.Command, args []string) {
+	ctx := context.Background()
+	var operation string
+	if len(args) > 0 {
+		operation = args[0]
+	}
+
+	vars := getKVInput(queryVarsInput)
+
+	// Use the provided query file if specified
+	// Otherwise, if stdin is a pipe or other non-tty thing, read from it.
+	// Finally, default to the operations returned by the loadExtension query
+	var operations string
+	if queryFile != "" {
+		inBytes, err := os.ReadFile(queryFile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+			os.Exit(1)
+		}
+		operations = string(inBytes)
+	} else if !term.IsTerminal(int(os.Stdin.Fd())) {
+		inBytes, err := io.ReadAll(os.Stdin)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+			os.Exit(1)
+		}
+		operations = string(inBytes)
+	}
+
+	result, err := doQuery(ctx, operations, operation, vars)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("%s\n", result)
+}
+
+func doQuery(ctx context.Context, query, op string, vars map[string]interface{}) ([]byte, error) {
+	opts := []dagger.ClientOpt{
+		dagger.WithWorkdir(workdir),
+		dagger.WithConfigPath(configPath),
+	}
+
+	if debugLogs {
+		opts = append(opts, dagger.WithLogOutput(os.Stderr))
+	}
+	c, err := dagger.Connect(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	defer c.Close()
+
+	res := make(map[string]interface{})
+	resp := &dagger.Response{Data: &res}
+	err = c.Do(ctx,
+		&dagger.Request{
+			Query:     query,
+			Variables: vars,
+			OpName:    op,
+		},
+		resp,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.Errors) > 0 {
+		return nil, resp.Errors
+	}
+
+	result, err := json.MarshalIndent(res, "", "    ")
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// TODO use generics?
+func getKVInput(kvs map[string]string) map[string]interface{} {
```

Comment:
Personally I think I'd rather go with a fallback option until we can implement this with full graphql type introspection. IMO there's just too many gotchas remaining even though I totally agree this covers the majority of cases (e.g. "true" is always coerced to be a bool, any string containing `,` is a list, etc.).

One fallback option would be to just have users provide vars in the form of a json blob, e.g.

```
dagger query --vars '{"foo": "bar", "baz": ["1", "true"]}' <<EOF 
{
 query Blah($foo: String!, $baz: [String!]!) {
     ...
  }
}
EOF
```

I realize that is a much worse ux but it seems like something we could start with and then layer the nicer `--set foo=bar` stuff on top once we have the implementation that uses graphql type introspection ready.

Just my opinion, not obvious at all. If we do decide to go this route then I think we should at least tighten up a few parts, such as only interpreting a bar as a list if it's wrapped in `[` and `]`.

Reply:
> One fallback option would be to just have users provide vars in the form of a json blob, e.g.

Yep, I agree that JSON blobs would be a much better fallback mechanism. Let's ask @shykes what he thinks about.